src/MPAGSCipher/ProcessCmdLine.cpp

```diff
@@ -0,0 +1,162 @@
+                    if(keySupplied){
+                        encrypt = true; //if key already supplied, fine
+                    } else {
+                        std::cerr << "[error] requires key argument to follow"  //if not and last element require key to be given
+                                << std::endl;
+                        return false;
+                    }
+                } else if (i != nCmdLineArgs - 1 and !keySupplied){
+                    for(std::size_t j{i};j<nCmdLineArgs;j++){   //if not the last element and no key supplied yet
+                        if(args[j]=="--key"){   //loop through all other remaining elements
+                            keySupplied = true; //if there is a key, make keySupplied element true and allow encryption
+                            encrypt = true;
+                            continue;   //stop looking after this
+                        } 
+                    }
+                    if(keySupplied){    //if keySupplied is true then good
+                        encrypt = true;
+                    } else {    //if we didnt find a key in the remaining elements of the arguments then give non zero return
+                        std::cerr << "[error] requires key argument to follow"
+                                << std::endl;
+                        return false;
+                    }
+                }
+            } else {    //makes sure that we can't use encrypt and decrypt simultaneously 
+                std::cerr << "[error] --encrypt and --decrypt cannot not be used simultaneously"
+                          << std::endl;
+                return false;
+            }
+        } else if(args[i] == "--decrypt") { //much like encrypt above but for decrypt 
+            if(!encrypt) {
+                if(i == nCmdLineArgs - 1) {
+                    if(keySupplied){
+                        decrypt = true;
+                    } else {
+                        std::cerr << "[error] requires key argument to follow"
+                                << std::endl;
+                        return false;
+                    }
+                } else {
+                    for(std::size_t j{i};j<nCmdLineArgs;j++){
+                        if(args[j]=="--key"){
+                            keySupplied = true;
+                            decrypt = true;
+                            continue;
+                        } 
+                    }
+                    if(keySupplied){
+                        decrypt = true;
+                    } else {
+                        std::cerr << "[error] requires key argument to follow"
+                                << std::endl;
+                        return false;
+                    }
+                }
+            } else {
+                std::cerr << "[error] --encrypt and --decrypt cannot not be used simultaneously"
+                            << std::endl;
+                return false;
+            }
```

Comment:
These two clauses for dealing with `encrypt` and `decrypt` are *very* complicated! Fundamentally, you just need to check if either flag is specified and if so set the appropriate variable. Deciding if you have or not at this point just overcomplicates (and potentially duplicates) the code. Also don't forget defaults are fine - if neither a key or encrypt/decrypt flag are specified, it's fine to assume the values (though you should print a warning in that case).